migrations/20221115082230_add_email_column_identities.up.sql

```diff
@@ -0,0 +1,12 @@
+update
+  auth.identities as identities
+set
+  identity_data = identity_data || jsonb_build_object('email', (select email from auth.users where id = identities.user_id))
+where identities.provider = 'email' and identity_data->>'email' = null;
+
+alter table only auth.identities
+  add column if not exists email text generated always as (lower(identity_data->>'email')) stored;
+
+comment on column auth.identities.email is 'Auth: Email is a generated column that references the optional email property in the identity_data';
+
+create unique index if not exists identities_email_provider_email_key on auth.identities (email) where (provider = 'email');
```

Comment:
i think we still need the unique index here to be on oauth providers too since we still have automatic linking right?

maybe this is safer till we remove it?
```suggestion
create unique index if not exists identities_email_provider_email_key on auth.identities (email) where (provider != 'saml/sso');
```

Reply:
Yeah you're absolutely right. I'll update the query!

Reply:
Updated. I had to add another index for querying, since the results from an analyze were thus:

Without index on `email`:

```
postgres=# explain analyze select * from auth.identities where provider = 'email' and email = 'abcdef';
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on identities  (cost=4.16..9.51 rows=1 width=168) (actual time=0.009..0.010 rows=0 loops=1)
   Recheck Cond: (provider = 'email'::text)
   Filter: (email = 'abcdef'::text)
   ->  Bitmap Index Scan on identities_pkey  (cost=0.00..4.16 rows=2 width=0) (actual time=0.005..0.005 rows=0 loops=1)
         Index Cond: (provider = 'email'::text)
 Planning Time: 1.154 ms
 Execution Time: 0.077 ms
(7 rows)
```

Conclustion: Does not use any index. My guess is that the `NOT LIKE` clause from the partial unique index is never "quick" enough for the query planner, so it just skips it.

With index on `email`:

```
postgres=# explain analyze select * from auth.identities where provider = 'email' and email = 'abcdef';
                                                         QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on identities  (cost=4.16..9.51 rows=1 width=168) (actual time=0.013..0.014 rows=0 loops=1)
   Recheck Cond: (email = 'abcdef'::text)
   Filter: (provider = 'email'::text)
   ->  Bitmap Index Scan on identities_email_idx  (cost=0.00..4.16 rows=2 width=0) (actual time=0.007..0.008 rows=0 loops=1)
         Index Cond: (email = 'abcdef'::text)
 Planning Time: 0.703 ms
 Execution Time: 0.075 ms
(7 rows)
```

Conclusion: Uses an index.